### ISSUE TITLE
[BUG]: Lib does not serialize correctly on optional (required = false) Array properties which can be undefined.

### ISSUE THREAD

**Issue by charbel-z** [GillianPerard/typescript-json-serializer#185]
### Version

4.0.1

### Description

I am using "typescript-json-serializer": "4.3.0" and I want to serialize a set of nested classes. It seems that the lib does not like 'required: false' Array properties that can be undefined -  if used lib doesn't work.

Actual Output 

`{"classB":[{"classC":{"myAttribute":true}}]}`

Correct Output

`{"classB":[{"classC":{"my_attribute":true}}]}`

I.e. ClassC property 'myAttribute' should be 'my_attribute' as specified in decorator 'name' option.

### Error

```Text
No Exception thrown
```

### Reproduction

```
import { JsonObject, JsonProperty, JsonSerializer } from 'typescript-json-serializer';

export type Optional<T> = T | undefined;

@JsonObject()
export class ClassC {
  @JsonProperty({ required: false, name: 'my_attribute' })
  public myAttribute: Optional<boolean>;

  constructor(myAttribute?: boolean) {
    this.myAttribute = myAttribute;
  }
}

@JsonObject()
export class ClassB {
  @JsonProperty({ required: true })
  public classC: ClassC;

  constructor(classC: ClassC) {
    this.classC = classC;
  }
}

@JsonObject()
export class ClassA {
  @JsonProperty({ required: false, type: ClassB })
  public classB: Array<ClassB> | undefined;

  constructor(classB?: Array<ClassB>) {
    this.classB = classB;
  }
}

describe('serialize nested classs', () => {
  it('serialize and honor custom property name', () => {
    const mapper = new JsonSerializer();

    const instanceC = new ClassC(true);

    const instanceB: ClassB = new ClassB(instanceC);

    const arr: Array<ClassB> = new Array(instanceB);
    const instanceA: ClassA = new ClassA(arr);

    const data = mapper.serialize(instanceA);

    console.log(JSON.stringify(data));

    expect(true).toEqual(true);
  });
});

```

### On which OS the bug appears?

macOS 11.6.8

### What is your project type?

NodeJS

### On which build mode the bug appears?

Development

**Comment by charbel-z** [comment]
It also looks like 'required' in ClassC is also not honoured. Unless I am using the lib wrongly?

**Comment by GillianPerard** [comment]
Hi thank you for using my lib.

The problem is that typescript loose the type if the property could be undefined.

I'm working on a solution.

**Comment by GillianPerard** [comment]
Problem is now fixed in the new 5.0.0 version!
